### PR TITLE
Extend fetchSize support across JPA Item Readers

### LINKED ISSUE #4479: Add `fetchSize` to `JpaCursorItemReaderBuilder`
**Expected Behavior**

`JpaCursorItemReaderBuilder` should have an option to set `fetchSize`, similar to `HibernateCursorItemReader` (which is deprecated) and `JdbcCursorItemReaderBuilder`.

**Current Behavior**

There is no option to fetch size for query executed by `JpaCursorItemReader`. As a workaround I created `FetchSizeAwareJpaQueryProvider`. Pasting here in case someone would like to use it:

```java
import jakarta.persistence.EntityManager;
import jakarta.persistence.Query;
import org.hibernate.jpa.AvailableHints;
import org.springframework.batch.item.database.orm.JpaQueryProvider;

/**
 * {@link JpaQueryProvider} implementation that enables setting "FETCH_SIZE" query hint on the JPA query.
 *
 * <a href="https://jdbc.postgresql.org/documentation/query/#getting-results-based-on-a-cursor">Getting results based on a cursor</a>
 * 
 * @author Maciej Walkowiak
 */
public class FetchSizeAwareJpaQueryProvider implements JpaQueryProvider {
    private final String queryString;
    private final int fetchSize;
    private EntityManager entityManager;

    public FetchSizeAwareJpaQueryProvider(String queryString, int fetchSize) {
        this.fetchSize = fetchSize;
        this.queryString = queryString;
    }

    @Override
    public Query createQuery() {
        Query query = this.entityManager.createQuery(queryString);
        query.setHint(AvailableHints.HINT_FETCH_SIZE, fetchSize);
        return query;
    }

    @Override
    public void setEntityManager(EntityManager entityManager) {
        this.entityManager = entityManager;
    }
}
```

.. and use it like this:

```java
new JpaCursorItemReaderBuilder<...>()
        .name("...")
        .entityManagerFactory(..)
        .queryProvider(new FetchSizeAwareJpaQueryProvider("<jpql query>", 100))
        .build();
```

If it makes sense I can see if I find time to contribute it.

### PR DESCRIPTION
**Pull request by baezzys**
Resolves #4479

### CODE REVIEW

File: spring-batch-infrastructure/src/main/java/org/springframework/batch/item/database/JpaCursorItemReader.java [hunk 1]
```diff
@@ -123,6 +137,9 @@ protected void doOpen() throws Exception {
 		if (this.parameterValues != null) {
 			this.parameterValues.forEach(query::setParameter);
 		}
+		if (this.fetchSize != 0) {
```
[the remark this answers is not in the document]
Comment (baezzys) [reply spring-batch-infrastructure/src/main/java/org/springframework/batch/item/database/JpaCursorItemReader.java]: @injae-kim I feel that using the default value of an int for validation is unclear. What are your thoughts on using a variable like VALUE_NOT_SET instead like AbstractCursorItemReader?

```
public static final int VALUE_NOT_SET = -1;

int fetch size = VALUE_NOT_SET;
```

[the remark this answers is not in the document]
Comment (injae-kim) [reply spring-batch-infrastructure/src/main/java/org/springframework/batch/item/database/JpaCursorItemReader.java]: Hmm I worry that maybe it's over engineering?🤔
Cause we can simply not use `FETCH_SIZE_HINT` when `fetchSize == 0`. (I think this way is simplest!)

```java
int fetchSize = -1;
...
afterPropertiesSet() {
  assert.state(fetchSize >= 0); // failed when user not set fetchSize!
}
```
Also, If we use `-1` as default fetch size then assertion failed when user not set fetchSize.
To fix this, logic will become more complicated. WDYT?

[the remark this answers is not in the document]
Comment (injae-kim) [reply spring-batch-infrastructure/src/main/java/org/springframework/batch/item/database/JpaCursorItemReader.java]: Thanks a lot! 👍

**Comment by fmbenhassine** [comment]
@baezzys Thank you for the PR! I added an inline comment on how we should address this feature request. Could please take a look? If you agree, please update the PR and it should be good to merge. Thank you upfront.

**Comment by baezzys** [comment]
@fmbenhassine Thanks for checking out my PR. I can't seem to find the inline comment you're talking about. Can you point me in the right direction?

**Review by ignite-jinwoo — APPROVED** [review]

**Review by ignite-jinwoo — COMMENTED** [review]
d

**Review by ignite-jinwoo — APPROVED** [review]

**Comment by fmbenhassine** [comment]
The JPA reader should not depend on APIs from a specific implementation (ie `import org.hibernate.jpa.AvailableHints;`), it should be implementation agnostic. The way the requested feature should be implemented is by adding a `Map<String, Object> hintValues` for hints (similar to `Map<String, Object> parameterValues` for query parameters) and set those hints on the query with `query#setHint` in `doOpen`. The user can then specify all the hints supported by the provider and pass them as a Map to the JPA reader.

**Comment by baezzys** [comment]
@fmbenhassine Thank you for your review and feedback! I'll update the PR to use Map<String, Object> hintValues for setting hints, ensuring the JPA reader remains implementation-agnostic.

**Comment by baezzys** [comment]
@fmbenhassine I've updated the PR to include a Map<String, Object> hintValues for hints, similar to our approach with query parameters. This change ensures our JPA reader is implementation-agnostic, allowing users to specify provider-supported hints directly.

I hope this update aligns with your guidance. Please let me know if there are any further adjustments needed.

**Comment by fmbenhassine** [comment]
Thank you for the update, LGTM now 👍 Rebased and merged as 5a261fa7c6945de9c3dc85145473c0fa3ac4321b. Thank you for your contribution!

**Comment by injae-kim** [comment]
Nice work @baezzys !

**Pull request closed without merging**